components/service/glean/src/main/java/mozilla/components/service/glean/storages/PingStorageEngine.kt

```diff
@@ -0,0 +1,178 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean.storages
+
+import android.content.Context
+import android.support.annotation.VisibleForTesting
+import kotlinx.coroutines.ObsoleteCoroutinesApi
+import kotlinx.coroutines.launch
+import mozilla.components.service.glean.Dispatchers
+import mozilla.components.service.glean.Glean
+import java.io.BufferedReader
+import java.io.BufferedWriter
+import java.io.File
+import java.io.FilenameFilter
+import java.io.FileNotFoundException
+import java.io.FileOutputStream
+import java.io.FileReader
+import java.io.IOException
+import java.io.OutputStreamWriter
+import java.util.regex.Matcher
+import java.util.regex.Pattern
+import java.util.UUID
+
+import mozilla.components.support.base.log.logger.Logger
+
+/**
+ * This class implementation stores pings as files on disk.
+ */
+internal open class PingStorageEngine(context: Context) {
+
+    private val logger: Logger = Logger("glean/PingStorageEngine")
+
+    @VisibleForTesting
+    internal val storageDirectory: File = File(context.applicationInfo.dataDir, PING_STORAGE_DIRECTORY)
+
+    companion object {
+        // Since ping file names are UUIDs, this matches UUIDs for filtering purposes
+        private const val FILE_PATTERN =
+            "[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}"
+        // Base directory for storing serialized pings
+        private const val PING_STORAGE_DIRECTORY = "${Glean.GLEAN_DATA_DIR}/ping_storage"
+    }
+
+    /**
+     * Function to store the ping to file for the [PingUploadWorker] to read and send when it is run
+     *
+     * @param pingType String representing the ping type: baseline, events, metrics (or custom ping)
+     * @param pingPath String representing the upload path used for uploading the ping
+     * @param ping Serialized JSON string representing the ping payload
+     */
+    @ObsoleteCoroutinesApi
+    @Synchronized
+    fun store(pingType: String, uuidFileName: UUID, pingPath: String, ping: String) {
+        Dispatchers.API.launch {
+            val pingStorageDirectory = File(storageDirectory, pingType)
+            assertDirectory(pingStorageDirectory)
+
+            // Build the file path for the ping, using the UUID from the path for the file name
+            val pingFile = File(pingStorageDirectory, uuidFileName.toString())
+
+            var stream: FileOutputStream? = null
+
+            try {
+                stream = FileOutputStream(pingFile, true)
+
+                val writer = BufferedWriter(OutputStreamWriter(stream))
+                writer.write(pingPath)
```

Comment:
Can we comment on what this is doing? Did you check with @pocmo why were they doing this and if there's any drawback?

Reply:
It's serializing the upload path on the first line of the file, and then it serializes the payload on the second line of the file.  I suppose it would be able to to recreate the path when it deserialized the ping, based on the filename and the ping's docType.  This was just how Telemetry lib did it so I followed suit.

Reply:
Sorry I was cryptic. I meant, can you *add* a comment explaining that in the code? :)